File: tree/Traversal/Postorder/iterative/python/PostorderTraversal.py
from collections.abc import Sequence


def PostorderTraversal(left_children: Sequence[int], right_children: Sequence[int], root: int) -> list[int]:
    order: list[int] = []
    stack: list[tuple[int, bool]] = [(root, False)]

    while stack:
        node, expanded = stack.pop()

        if expanded:
            order.append(node)
            continue

        stack.append((node, True))

        right_child = right_children[node]
        if right_child != -1:
            stack.append((right_child, False))

        left_child = left_children[node]
        if left_child != -1:
            stack.append((left_child, False))

    return order

File: tree/Traversal/Postorder/iterative/java/PostorderTraversal.java
import java.util.*;

public class PostorderTraversal {
    private static final class Frame {
        final int node;
        final boolean expanded;

        Frame(int node, boolean expanded) {
            this.node = node;
            this.expanded = expanded;
        }
    }

    public static List<Integer> PostorderTraversal(int[] leftChildren, int[] rightChildren, int root) {
        List<Integer> order = new ArrayList<>();
        Deque<Frame> stack = new ArrayDeque<>();
        stack.push(new Frame(root, false));

        while (!stack.isEmpty()) {
            Frame frame = stack.pop();
            if (frame.expanded) {
                order.add(frame.node);
                continue;
            }

            stack.push(new Frame(frame.node, true));

            int rightChild = rightChildren[frame.node];
            if (rightChild != -1) {
                stack.push(new Frame(rightChild, false));
            }

            int leftChild = leftChildren[frame.node];
            if (leftChild != -1) {
                stack.push(new Frame(leftChild, false));
            }
        }

        return order;
    }
}
